test/controllers/observer_controller_test.rb

```diff
@@ -2880,6 +3027,76 @@ def test_edit_observation_strip_images
     assert_false(old_img2.reload.gps_stripped)
   end
 
+  def test_edit_observation_save_failure
+    obs = observations(:minimal_unknown_obs)
+    params = {
+      id: obs.id.to_s,
+      observation: {
+        notes: { other: "new notes" },
+        place_name: obs.where,
+        "when(1i)" => "2006",
+        "when(2i)" => "05",
+        "when(3i)" => "11",
+        specimen: obs.specimen,
+        thumb_image_id: obs.thumb_image_id
+      },
+      log_change: { checked: "0" }
+    }
+
+    login(obs.user.name)
+    Observation.any_instance.stubs(:save).returns(false)
+    assert_no_difference(
+      "Observation.count", "An Observation should not be created"
+    ) do
+      post(:edit_observation, params: params)
+    end
+
+    assert_flash_text(/#{:runtime_no_save.l(type: "observation")}/)
+    assert_response(:success, "Edit form should be reloaded")
+  end
+
+  def test_edit_observation_image_save_failure
+    obs = observations(:detailed_unknown_obs)
+    # more detailed location to avoid location flash warning and redirection
+    location = locations(:point_reyes)
+    obs.update(location: location)
+    img = obs.images.first
+    img_notes = img.notes
+
+    params = {
+      id: obs.id.to_s,
+      observation: {
+        notes: obs.notes,
+        place_name: obs.where,
+        "when(1i)" => obs.when.strftime("%Y"),
+        "when(2i)" => obs.when.strftime("%m"),
+        "when(3i)" => obs.when.strftime("%d"),
+        specimen: obs.specimen,
+        thumb_image_id: obs.thumb_image_id
+      },
+      good_images: "#{img.id} #{images(:turned_over_image).id}",
+      good_image: {
+        img.id.to_s => {
+          notes: "change something to force image to be saved",
+          original_name: img.original_name,
+          copyright_holder: img.copyright_holder,
+          "when(1i)" => img.when.strftime("%Y"),
+          "when(2i)" => img.when.strftime("%m"),
+          "when(3i)" => img.when.strftime("%d"),
+          license_id: img.license_id
+        }
+      },
+      log_change: { checked: "0" }
+    }
+
+    login(obs.user.name)
+    Image.any_instance.stubs(:save).returns(false)
+    post(:edit_observation, params: params)
+
+    assert_redirected_to(/#{obs.id}$/)
+    assert_equal(img_notes, img.reload.notes)
```

Comment:
There is also a problem with my test (or maybe with the code being tested).
While I am covering the line, the assertion isn't what I want: 
I'd like to assert that the flash has an Image error. But when I look at the flash here, there's no Image error. 
(And I think the assertion I wrote is actually testing whether `stubs`. So I probably should get rid of it.)